src/stream_insert.rs

```diff
@@ -0,0 +1,115 @@
+// limitations under the License.
+
+use greptime_proto::v1::greptime_request::Request;
+use greptime_proto::v1::{greptime_database_client::GreptimeDatabaseClient, InsertRequest};
+use greptime_proto::v1::{
+    greptime_response, AffectedRows, AuthHeader, GreptimeRequest, GreptimeResponse, InsertRequests,
+    RequestHeader,
+};
+use snafu::OptionExt;
+use tokio::sync::mpsc;
+use tokio::task::JoinHandle;
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::transport::Channel;
+use tonic::{Response, Status};
+
+use crate::error::Result;
+use crate::error::{self, IllegalDatabaseResponseSnafu};
+
+/// A structure that provides some methods for streaming data insert.
+///
+/// [`StreamInsertor`] cannot be constructed via the `StreamInsertor::new` method.
+/// You can use the following way to obtain [StreamInsertor].
+///
+/// ```ignore
+/// let grpc_client = Client::with_urls(vec!["127.0.0.1:4002"]);
+/// let client = Database::new_with_dbname("db_name", grpc_client);
+/// let stream_insertor = client.streaming_insertor().unwrap();
+/// ```
+///
+/// If you want to see a concrete usage example, please see
+/// [stream_insertor.rs](https://github.com/GreptimeTeam/greptimedb-client-rust/tree/master/examples/stream_ingest.rs).
+pub struct StreamInsertor {
+    sender: mpsc::Sender<GreptimeRequest>,
+
+    auth_header: Option<AuthHeader>,
+
+    dbname: String,
+
+    join: JoinHandle<std::result::Result<Response<GreptimeResponse>, Status>>,
+}
+
+impl StreamInsertor {
+    pub(crate) fn new(
+        mut client: GreptimeDatabaseClient<Channel>,
+        dbname: String,
+        auth_header: Option<AuthHeader>,
+    ) -> StreamInsertor {
+        let (send, recv) = tokio::sync::mpsc::channel(1024);
+
+        let join: JoinHandle<std::result::Result<Response<GreptimeResponse>, Status>> =
+            tokio::spawn(async move {
+                let recv_stream = ReceiverStream::new(recv);
+                client.handle_requests(recv_stream).await
+            });
+
+        StreamInsertor {
+            sender: send,
+            auth_header,
+            dbname,
+            join,
+        }
+    }
+
+    pub async fn insert(&self, requests: Vec<InsertRequest>) -> Result<()> {
+        let inserts = InsertRequests { inserts: requests };
+        let request = self.to_rpc_request(Request::Inserts(inserts));
+
+        self.sender.send(request).await.map_err(|e| {
+            error::ClientStreamingSnafu {
+                err_msg: e.to_string(),
+            }
+            .build()
+        })
+    }
+
+    pub async fn finish(self) -> Result<u32> {
+        drop(self.sender);
+
+        let response = self.join.await.unwrap()?;
+
+        let response = response
+            .into_inner()
+            .response
+            .context(IllegalDatabaseResponseSnafu {
+                err_msg: "GreptimeResponse is empty",
+            })?;
+
+        let greptime_response::Response::AffectedRows(AffectedRows { value }) = response;
+
+        Ok(value)
```

Comment:
I tend to directly return `greptime_response::Response`

Reply:
I see that Result\<u32\> is returned [here](https://github.com/GreptimeTeam/greptimedb-client-rust/blob/master/src/database.rs#L72), shouldn’t it be consistent with this?

Reply:
Let's keep the same style, using `Result<u32>`.